sdk/transport_policies/curl/src/az_curl.c

```diff
@@ -314,6 +316,7 @@ _az_http_client_curl_send_upload_request(CURL * p_curl, _az_http_request const *
 
   char * b = (char *)az_span_ptr(body);
```

Comment:
it looks like this method `upload` was trying to imitate what method `post` does.
And `post` is imitating what `set_url` does.
But this is different completely.  For `set_url` we need to allocate a new buffer in memory because we need it to be 0-terminated.
I think for posting a body, or uploading a body, we shouldn't need to add a 0 at the end of it (because it changes the actual body of it).  
It should be enough to use the body span directly.
Like 

```c
static AZ_NODISCARD az_result
_az_http_client_curl_send_upload_request(CURL * p_curl, _az_http_request const * p_request) {
  AZ_CONTRACT_ARG_NOT_NULL(p_curl);
  AZ_CONTRACT_ARG_NOT_NULL(p_request);

  // Method
  az_span body = p_request->_internal.body;

  AZ_RETURN_IF_CURL_FAILED(curl_easy_setopt(p_curl, CURLOPT_UPLOAD, 1L));

  AZ_RETURN_IF_CURL_FAILED(
      curl_easy_setopt(p_curl, CURLOPT_READFUNCTION, _az_http_client_curl_upload_read_callback));

  AZ_RETURN_IF_CURL_FAILED(curl_easy_setopt(p_curl, CURLOPT_READDATA, body));

  AZ_RETURN_IF_CURL_FAILED(curl_easy_setopt(p_curl, CURLOPT_INFILESIZE, az_span_length(body)));

  // Do the curl work
  return _az_http_client_curl_code_to_result(curl_easy_perform(p_curl));
}
```

Reply:
Actually `post` method does need to allocate because it uses `CURLOPT_POSTFIELDS`  and it is expected to be 0-terminated. : (https://curl.haxx.se/libcurl/c/CURLOPT_POSTFIELDS.html).